### PR TITLE
feat(blocks-engine): ask whether a document's component references could all be read

### PR DESCRIPTION
**Pull request by mobeenabdullah**
First slice of Plan 06 **T-5**, the component usage index. This is the
primitive the index rests on; no storage and no hook yet.

## The problem

`componentIdsIn` walks a bounded forest and **stops silently**. A document too
large to read whole therefore returns the same `[]` as a document that
references no components at all.

That is the wrong way round for the question a usage index asks. The index
exists to decide whether a component may be deleted, and a delete check reads
absence as evidence — so an absence produced by a *bound* rather than by the
document is exactly the answer that permits deleting a component a live page
still renders.

The class usage index already solved this for classes: `classUsageOf` returns
`{ ids, complete }` for precisely this reason. This gives components the same
two-valued answer.

## The shape

`componentUsageIn(nodes, maxNodes) → { ids, complete }`, and **`componentIdsIn`
is now derived from it** rather than walking beside it. They are one question,
and two walks over one forest agree on the day they are written — the narrow
one goes on returning a prefix silently once a bound, an ordering or a descent
rule moves only the other.

Order stays first-reached rather than sorted, because `componentIdsIn`
publishes that and its consumer is cache tagging.

## The boundary case, and why it has its own test

`complete` is set on the branch that **ends** the walk, not derived afterwards
from the leftover budget. A forest holding exactly `maxNodes` entries spends
the last of the budget on its last entry and is read **whole** — so
`complete: budget > 0` would report a complete read as truncated, at exactly
the size the rest of the engine still accepts.

## Break-verification

Three wrong implementations, each killing a **different** test:

| wrong implementation | test that fails |
|---|---|
| `complete` derived from the leftover budget | `calls a forest of exactly the budget COMPLETE, not truncated` |
| never records truncation | `reports a truncated read, and the prefix it managed` |
| `componentIdsIn` walks under its own budget | `answers what componentIdsIn answers, truncation included`, and the pre-existing budget test |

One caveat stated rather than glossed: the second break's edit matched in two
places (`truncated = true` occurs twice in the file), so it was broader than a
single-site mutation. The target test still failed, but it was not surgical.

The truncated-case test asserts `ids` **and** `complete` together, deliberately:
`ids: []` alone is also what a document holding no instances returns, so an
assertion on it cannot separate the two states this function exists to separate.

## Gates

| gate | result |
|---|---|
| `blocks-engine` `check-types` | exit 0 (both `tsc` invocations) |
| `blocks-engine` `lint` | exit 0 |
| `blocks-engine` tests | 52 files, 2411 tests, exit 0 |
| `blocks-react` tests (names `componentIdsIn`) | 48 files, 1162 tests, exit 0 |
| `builder` tests | 102 files, 2869 tests, exit 0 |
| `fallow audit --base origin/main` | `pass`, `changed_files_count: 4`, introduced 0 across all four categories |
| `pnpm changeset status` | parses; contributes 25 packages, verified by diffing status with and without the file |

## Summary by CodeRabbit

- **New Features**
  - Added a component usage API that returns referenced component IDs and indicates whether traversal completed or was limited by the configured budget.
  - Exposed the new usage result type for consumers integrating with the blocks engine.
  - Existing component ID retrieval continues to provide the same array-based results while using the enhanced traversal information internally.
- **Tests**
  - Added coverage for complete and truncated traversal, budget boundaries, empty inputs, and consistency with existing component ID results.

### CODE REVIEW

File: packages/blocks-engine/src/resolve-instances.ts [hunk 1]
```diff
@@ -262,16 +262,58 @@ const SCOPED_ID_PREFIX = "cx-";
...
+ *
+ * ORDER IS FIRST-REACHED, not sorted, because `componentIdsIn` publishes that
+ * and its consumer is cache tagging. A caller comparing two answers for
+ * equality has to sort them itself.
+ */
+export function componentUsageIn(
   nodes: readonly unknown[],
   maxNodes: number = DEFAULT_LIMITS.maxNodes
-): string[] {
+): ComponentUsage {
```
[the remark this answers is not in the document]
Comment (mobeenabdullah) [reply packages/blocks-engine/src/resolve-instances.ts:302]: Reproduced both halves, and fixed the one that matters — thank you.

Measured on three entries:

| `maxNodes` | ids | complete | |
|---|---|---|---|
| `3` | `["late"]` | `true` | correct |
| `2` | `[]` | `false` | correct |
| `2.5` | `["late"]` | `true` | reads 3 — one over |
| `NaN` | `["late"]` | `true` | **no bound at all** |

`NaN` is the serious one, and worse than "exceeds the budget": every
`budget <= 0` test against it is false, so the walk has no stop condition and
reads a stored document of any size whole. That is now refused.

**Guarded through `boundedLimit` rather than the proposed inline check.** That
helper already exists in this package for exactly this, and its docblock says
why it is published: *"more than one walk is held to these numbers and they
have to agree about what counts as a bound."* A second rule spelled locally is
how two walks come to accept different budgets — which is the same failure this
PR is about, one level up. It also already refuses non-numbers, which
`Number.isSafeInteger` alone would have covered differently, and deliberately
permits `Infinity` as the supported way to ask for an exact count.

**Fractional budgets stay accepted, deliberately.** Refusing them here and
nowhere else would make this walk disagree with the helper that exists to stop
walks disagreeing — and the cost is bounded at one node past the cap, with the
answer still truthful, because `complete` reports what the walk *reached*
rather than what the budget *allowed*. There is a test pinning both directions
(`2.5` → complete, `1.5` → truncated), so a rule that only ever reported
`complete: true` would not satisfy it.

`componentIdsIn` inherits the refusal, which is what being derived means — its
previous behaviour on `NaN` was to walk unbounded in silence. That is a
behaviour change to a published function, so the changeset says so.

Two notes on the verification, stated rather than glossed:

- Removing the guard fails the new test, so it is break-verified.
- Swapping the validated value for the raw one **kills nothing**, because
  `boundedLimit` returns its input unchanged and the two are the same number on
  every path past it. I kept the returned value anyway — a helper that later
  normalises would be honoured for free — but the code now says outright that no
  test can tell the two spellings apart, so it does not read as covered.

Separately, `surveyHost` in the same file has the identical unbounded-`NaN`
shape and is reached from the published `resolveComponentInstances`. It is
pre-existing and changing it would make that function start throwing on a path
this PR does not otherwise touch, so I have recorded it rather than widened this
change.

**Comment by mobeenabdullah** [comment]
@codex review
